metricbeat/module/consul/agent/data.go

```diff
@@ -0,0 +1,100 @@
+package agent
+
+import (
+	"encoding/json"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/safemapstr"
+	"strings"
+)
+
+type Gauge struct {
+	Name   string            `json:"Name"`
+	Value  float64           `json:"Value"`
+	Labels map[string]string `json:"Labels"`
+}
+
+type Counter consulDetailedValue
+
+type Sample consulDetailedValue
+
+type consulDetailedValue struct {
+	Name   string            `json:"Name"`
+	Count  int               `json:"Count"`
+	Rate   float64           `json:"Rate"`
+	Sum    float64           `json:"Sum"`
+	Min    float64           `json:"Min"`
+	Max    float64           `json:"Max"`
+	Mean   float64           `json:"Mean"`
+	Stddev float64           `json:"Stddev"`
+	Labels map[string]string `json:"Labels"`
+}
+
+// TODO Point has been copied from Gauge. REVIEW and place the exact type it needs.
+type Point struct {
+	Name   string            `json:"Name"`
+	Value  float64           `json:"Value"`
+	Labels map[string]string `json:"Labels"`
+}
+
+type Agent struct {
+	Timestamp string    `json:"Timestamp"`
+	Gauges    []Gauge   `json:"Gauges"`
+	Points    []Point   `json:"Points"`
+	Counters  []Counter `json:"Counters"`
+	Samples   []Sample  `json:"Samples"`
+}
+
+func eventMapping(content []byte) (common.MapStr, error) {
+	var agent Agent
+
+	if err := json.Unmarshal(content, &agent); err != nil {
+		return nil, err
+	}
+
+	event := common.MapStr{"timestamp": agent.Timestamp}
+
+	for _, gauge := range agent.Gauges {
+		safemapstr.Put(event, removeConsulFromBeginningOfName(gauge.Name), common.MapStr{
+			"value":  gauge.Value,
+			"labels": gauge.Labels,
+		})
+	}
+
+	for _, point := range agent.Points {
+		safemapstr.Put(event, removeConsulFromBeginningOfName(point.Name), common.MapStr{
+			"value":  point.Value,
+			"labels": point.Labels,
+		})
+	}
+
+	for _, counter := range agent.Counters {
+		safemapstr.Put(event, removeConsulFromBeginningOfName(counter.Name), common.MapStr{
+			"value":  counter.Count,
+			"labels": counter.Labels,
+		})
+	}
+
+	for _, sample := range agent.Samples {
+		sampleMetric := fillConsulDetailedValue(consulDetailedValue(sample))
+		safemapstr.Put(event, removeConsulFromBeginningOfName(sample.Name), sampleMetric)
+	}
+
+	return event, nil
+}
+
+func removeConsulFromBeginningOfName(s string) string {
+	return strings.ToLower(strings.Replace(s, "consul.", "", -1))
+}
+
+func fillConsulDetailedValue(v consulDetailedValue) common.MapStr {
+	return common.MapStr{
+		"count":  v.Count,
+		"rate":   v.Rate,
+		"sum":    v.Sum,
+		"min":    v.Min,
+		"max":    v.Max,
+		"mean":   v.Mean,
+		"stddev": v.Stddev,
+		"labels": v.Labels,
+	}
```

Comment:
You will need to add a dynamic template so all `consul.*.rate` and so on fields are defined as float.